ngraph/core/src/op/matmul.cpp

```diff
@@ -113,68 +53,152 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
+        std::vector<Dimension> output_shape(big_batch_matrix);
+
+        if (arg0_rank != arg1_rank)
         {
-            // i.e., arg0 shape {2, 2, 3}, arg1 shape{3}, output shape {2, 2}
-            NGRAPH_CHECK(arg1_shape_update[0] == arg0_shape_update[arg0_rank - 1],
-                         "Incompatible arg shapes");
-            arg0_shape_update.erase(arg0_shape_update.begin() + arg0_rank - 1);
-            output_shape = arg0_shape_update;
+            // Expand small_batch_matrix (with 1) to have the same rank as big_batch_matrix
+            size_t delta_rank = big_batch_matrix.size() - small_batch_matrix.size();
+            small_batch_matrix.insert(small_batch_matrix.begin(), delta_rank, 1);
         }
-        else if (arg0_rank == 2 && arg1_rank == 2)
+
+        // 4. Usual rules of the broadcasting are applied for batch dimensions.
+        // Broadcast all batches (last two dimensions represent matrix),
+        // expand dim with value 1 to bigger dim if dimensions are not equal.
+        for (auto i = 0; i < big_batch_matrix.size() - 2; i++)
         {
-            NGRAPH_CHECK(arg0_shape_update[1] == arg1_shape_update[0], "Incompatible arg shapes");
-            output_shape = Shape{arg0_shape_update[0], arg1_shape_update[1]};
+            // Dimension with value 1 can be expanded to any bigger
+            auto min_dim_val = std::min(small_batch_matrix[i].get_min_length(),
+                                        big_batch_matrix[i].get_min_length());
+            if (min_dim_val > 1)
+            {
+                auto merged_dimension = Dimension::dynamic();
+                NGRAPH_CHECK(Dimension::merge(
+                                 merged_dimension, small_batch_matrix[i], big_batch_matrix[i]) ||
+                                 small_batch_matrix[i].is_dynamic() ||
+                                 big_batch_matrix[i].is_dynamic(),
+                             "Incompatible MatMul batch dimension. ",
+                             "Can't merge dimension=",
+                             small_batch_matrix[i],
+                             " with dimension=",
+                             big_batch_matrix[i],
+                             " at index=",
+                             i);
+
+                output_shape[i] = merged_dimension;
+            }
+            else
+            {
+                Dimension::value_type upper_bound, lower_bound;
+                lower_bound = std::max(small_batch_matrix[i].get_min_length(),
+                                       big_batch_matrix[i].get_min_length());
+
+                if (lower_bound <= 1)
+                {
+                    // Both of the dimensions have 1 in range,
+                    // upper_bound is the maximum of the each range highest possible value.
+                    upper_bound = std::max(small_batch_matrix[i].get_interval().get_max_val(),
+                                           big_batch_matrix[i].get_interval().get_max_val());
```

Comment:
Use `get_max_length`.

Reply:
[Dimension::get_max_length()](https://github.com/openvinotoolkit/openvino/blob/9c509e5f416cce29dddd3bd1de292055df3765c5/ngraph/core/src/dimension.cpp#L140) will return `-1` for the upper bound limit `Interval::s_max`. 
So `std::max` for `-1` as upper bound limit won't be correct.
I use `dimension.get_interval().get_max_val()` to get `Interval::s_max` value instead of `-1`.

Reply:
You are right. This is an artifact (really an issue) of PartialShape that I've missed.